graph/telemetry/istio/appender/service_entry.go

```diff
@@ -113,19 +113,50 @@ func (a ServiceEntryAppender) applyServiceEntries(trafficMap graph.TrafficMap, g
 			for _, doomedEdge := range doomedSeServiceNode.Edges {
 				var aggregateEdge *graph.Edge
 				for _, e := range serviceEntryNode.Edges {
-					if doomedEdge.Dest.ID == e.Dest.ID {
+					if doomedEdge.Dest.ID == e.Dest.ID && doomedEdge.Metadata[graph.ProtocolKey] == e.Metadata[graph.ProtocolKey] {
 						aggregateEdge = e
 						break
 					}
 				}
 				if nil == aggregateEdge {
 					aggregateEdge = serviceEntryNode.AddEdge(doomedEdge.Dest)
+					aggregateEdge.Metadata[graph.ProtocolKey] = doomedEdge.Metadata[graph.ProtocolKey]
 				}
 				graph.AggregateEdgeTraffic(doomedEdge, aggregateEdge)
 			}
 			delete(trafficMap, doomedSeServiceNode.ID)
 		}
 		trafficMap[serviceEntryNode.ID] = &serviceEntryNode
+
+		// If there is more than one doomed node, edges leading to the new aggregated node must
+		// also be aggregated per source and protocol.
+		if len(seServiceNodes) > 1 {
+			for _, n := range trafficMap {
+				// Get the edges that need to be aggregated and remove them from the node.
+				edgesToAggregate := make(map[string][]*graph.Edge) // This is per protocol.
+				bound := 0
+				for _, edge := range n.Edges {
+					if edge.Dest == &serviceEntryNode {
```

Comment:
Maybe this block could be made into a function and called above after line 110, where we already loop through the nodes.  That is where we are creating the edges that need to be aggregated I think.

Reply:
OK, I extracted this code to avoid making this function larger.

About the placement, I have the idea that aggregating edges here, results in less iterations because it's out of the `for` over `seServiceNodes`. But I don't think this will result in any performance impact, so I'm moving the code where you suggested.

Reply:
I would have been fine with or without the change.  There may be more iterations but I also don't think there is a perf impact.  I like now that the aggregation logic is in a function, and I think the code is easier to understand because you're checking for the issue when actively looking at a potentially problematic source node.  So thanks for the change.